src/meta/src/barrier/backfill_order_control.rs

```diff
@@ -0,0 +1,149 @@
+// Copyright 2025 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+
+use risingwave_pb::stream_plan::FragmentTypeFlag;
+
+use crate::model::{FragmentId, StreamJobFragments};
+
+/// This is the "global" `fragment_id`.
+/// The local `fragment_id` is namespaced by the `fragment_id`.
+pub type ActorId = u32;
+
+#[derive(Clone, Debug, Default)]
+pub struct BackfillNode {
+    fragment_id: FragmentId,
+    /// How many more actors need to finish,
+    /// before this fragment can finish backfilling.
+    remaining_actor_count: usize,
+    /// How many more dependencies need to finish,
+    /// before this fragment can be backfilled.
+    remaining_dependency_count: usize,
+    children: Vec<FragmentId>,
+}
+
+/// Actor done                   -> update `fragment_id` state
+/// Operator done                -> update downstream operator dependency
+/// Operator's dependencies done -> queue operator for backfill
+#[derive(Clone, Debug)]
+pub struct BackfillOrderState {
+    // The order plan.
+    current_backfill_nodes: HashMap<FragmentId, BackfillNode>,
+    // Remaining nodes to finish
+    remaining_backfill_nodes: HashMap<FragmentId, BackfillNode>,
+    // The mapping between actors and fragment_ids
+    actor_to_fragment_id: HashMap<ActorId, FragmentId>,
+}
+
+// constructor
+impl BackfillOrderState {
```

Comment:
What if users accidentally specify a cycle and we cannot do topological sort here?

Reply:
Yeah I previously thought of this as well, and mentioned it in the future work section (Cycle detection).

The PR is large at the moment. I will refine this later.

Reply:
https://github.com/risingwavelabs/risingwave/pull/20967/commits/9b4ac0c570a87b5d4a93e863a1e9376e7752e11e

Reply:
Added